bridges/otelzap/zap_logger.go

```diff
@@ -0,0 +1,119 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+// Package otelzap provides a bridge between the ["go.uber.org/zap"] and
+// OpenTelemetry logging.
+package otelzap // import "go.opentelemetry.io/contrib/bridges/otelzap"
+
+import (
+	"context"
+
+	"go.uber.org/zap/zapcore"
+
+	"go.opentelemetry.io/otel/log"
+)
+
+const (
+	bridgeName = "go.opentelemetry.io/contrib/bridge/zapcore"
+)
+
+// Core is a [zapcore.Core] that sends all logging records it receives to
+// OpenTelemetry.
+type Core struct {
+	logger log.Logger
+	attr   []log.KeyValue
+}
+
+// Compile-time check *Core implements zapcore.Core.
+var _ zapcore.Core = (*Core)(nil)
+
+// this function creates a new zapcore.Core that can be used with zap.New()
+// this instance will translate zap logs to opentelemetry logs and export them.
+func NewOtelZapCore(opts ...Option) zapcore.Core {
+	cfg := newConfig(opts)
+	return &Core{
+		logger: cfg.logger(),
+	}
+}
+
+// LevelEnabler decides whether a given logging level is enabled when logging a
+// message.
+func (o *Core) Enabled(level zapcore.Level) bool {
+	r := log.Record{}
+	r.SetSeverity(getOtelLevel(level))
+
+	// TODO:
+	// check how to propagate context
+	ctx := context.Background()
+	return o.logger.Enabled(ctx, r)
+}
+
+// With adds structured context to the Core.
+// by returning a child logger with provided fields.
+func (o *Core) With(fields []zapcore.Field) zapcore.Core {
+	clone := o.clone()
+	clone.attr = append(clone.attr, getAttr(fields)...)
+	return clone
+}
+
+// Sync flushes buffered logs (if any).
+func (o *Core) Sync() error {
+	return nil
+}
+
+// Check determines whether the supplied Entry should be logged using core.Enabled method.
+func (o *Core) Check(ent zapcore.Entry, ce *zapcore.CheckedEntry) *zapcore.CheckedEntry {
+	if o.Enabled(ent.Level) {
+		return ce.AddCore(ent, o)
+	}
+	return ce
+}
+
+// Writes to the destination.
+func (o *Core) Write(ent zapcore.Entry, fields []zapcore.Field) error {
+	r := log.Record{}
+	r.SetTimestamp(ent.Time)
+	r.SetBody(log.StringValue(ent.Message))
+	r.SetSeverity(getOtelLevel(ent.Level))
+
+	// get attr from fields
+	attr := getAttr(fields)
+	// append attributes received from from parent logger
+	addattr := append(attr, o.attr...)
+
+	if len(addattr) > 0 {
+		r.AddAttributes(addattr...)
+	}
+
+	// TODO
+	// need to check how to propagate context here
```

Comment:
I think that if one passed a `context.Context` via attribute using https://pkg.go.dev/go.uber.org/zap#Any e.g.

```
logger.Info("failed to fetch URL",
  zap.Any("context", ctx),
  zap.Int("attempt", 3),
  zap.Duration("backoff", time.Second),
)
```

then we can use it.

If have done something like this in logr bridge prototype https://github.com/pellared/opentelemetry-go/blob/147762ffb5c4394417b3f973274cc570723db79b/log/internal/logr.go#L49-L53

This behavior should be clearly documented in `otelzap` package documentation.

Reply:
E.g. 



This is how the bridge can get the passed context:

```go
// converts zap fields to Otel's log attributes.
func getAttr(ctx context.Context, fields []zapcore.Field) (context.Context, []log.KeyValue) {
	enc := NewObjectEncoder(len(fields))
	for _, fld := range fields {
		if ctxFld, ok := fld.Interface.(context.Context); ok {
			ctx = ctxFld
			continue
		}
		fld.AddTo(enc)
	}
	return ctx, enc.cur
}
```